lib/rules/string-no-newline/index.mjs

```diff
@@ -56,67 +55,54 @@ const rule = (primary) => {
 				return;
 			}
 
-			parseSelector(ruleNode.selector, result, ruleNode)?.walkAttributes((attributeNode) => {
-				const { value, quoteMark } = attributeNode;
-
-				if (!value || !reNewLine.test(value)) {
-					return;
-				}
-
-				const openIndex = [
-					// length of our attribute
-					attributeNode.attribute,
-					// length of our operator , ie '='
-					attributeNode.operator || '',
-				].reduce(
-					(index, str) => index + str.length,
-					// index of the start of our attribute node in our source
-					// plus 1 for the opening quotation mark
-					attributeNode.sourceIndex + 1,
-				);
-
-				const valueLength = value.length + (quoteMark || '').length * 2;
-
-				report({
-					message: messages.rejected,
-					node: ruleNode,
-					index: openIndex,
-					endIndex: openIndex + valueLength,
-					result,
-					ruleName,
-				});
-			});
+			check(ruleNode, ruleNode.selector, () => 0);
 		}
 
 		/**
-		 * @template {import('postcss').AtRule | import('postcss').Declaration} T
+		 * @template {import('postcss').AtRule | import('postcss').Rule | import('postcss').Declaration} T
 		 * @param {T} node
 		 * @param {string} value
 		 * @param {(node: T) => number} getIndex
 		 * @returns {void}
 		 */
-		function checkDeclOrAtRule(node, value, getIndex) {
+		function check(node, value, getIndex) {
 			// Get out quickly if there are no new line
 			if (!reNewLine.test(value)) {
 				return;
 			}
 
-			valueParser(value).walk((valueNode) => {
-				if (valueNode.type !== 'string') {
-					return;
-				}
+			/** @typedef {import('@csstools/css-tokenizer').ParseError} ParseError */
+			/** @type {Array<ParseError>} */
+			let errors = [];
+			const tokens = tokenize(
+				{ css: value },
+				{
+					onParseError: (/** @type {ParseError} */ err) => {
+						if (err.message === 'Unexpected newline while consuming a string token.') {
+							errors.push(err);
+						}
+					},
```

Comment:
This rule happens to align perfectly with a specified parse error: https://drafts.csswg.org/css-syntax/#consume-string-token

> newline
> This is a parse error. Reconsume the current input code point, create a `<bad-string-token>`, and return it.